src/state/signalr-connection.ts

```diff
@@ -0,0 +1,139 @@
+import _Vue from 'vue';
+import { LogLevel, HubConnection, HubConnectionBuilder } from '@microsoft/signalr';
+import { StateManager } from '@/state';
+import { SignalRUtilities, NotificationHandler } from '@/dtos/temp-signalr';
+
+type ConnectionStatus = 'closed' | 'connecting' | 'connected' | 'closing';
+
+export class SignalRWrapper {
+    // Manage the SignalR connection. This is a singleton class
+    private static _instance: SignalRWrapper;
+
+    private _connection?: HubConnection;
+    private _utils?: SignalRUtilities;
+    private _currentHandler?: NotificationHandler;
+    private _status: ConnectionStatus = 'closed';
+    private _subscribedEditionId?: number;
+
+    public static get instance() {
+        if (!SignalRWrapper._instance) {
+            SignalRWrapper._instance = new SignalRWrapper();
+        }
+
+        return SignalRWrapper._instance;
+    }
+
+    private constructor() {
+        if (!process.env.VUE_APP_SIGNALR_URL) {
+            throw new Error('VUE_APP_SIGNALR_URL not defined, please add it to the environment');
+        }
+    }
+
+    public async subscribeEdition(editionId: number) {
+        console.debug('SignalR subscribeEdition called, current status ', this._status);
+        if (this._subscribedEditionId === editionId) {
+            return;
+        }
+
+        if (this._status !== 'connected') {
+            await this.connect();
+        }
+
+        if (this._subscribedEditionId) {
+            await this.unsubscribeEdition();
+        }
+
+        await this._utils!.subscribeToEdition(editionId);
+        console.debug('SubscribeToEdition called');
+        this._subscribedEditionId = editionId;
+    }
+
+    public async unsubscribeEdition() {
+        console.debug('SignalR unsubscribeEdition, status is ', this._status);
+        if (this._status !== 'connected' || !this._subscribedEditionId) {
+            return;
+        }
+
+        await this._utils!.unsubscribeToEdition(this._subscribedEditionId);
+        console.debug('UnsubscribeToEdition called');
+        this._subscribedEditionId = undefined;
+    }
+
+    public async userChanged() {
+        console.debug('SignalR changing users');
+        const subscribed = this._subscribedEditionId;
+        await this.disconnect();
+        await this.connect();
+        if (subscribed) {
+            await this.subscribeEdition(subscribed);
+        }
+    }
+
+    public registerNotificationHandler(handler: NotificationHandler) {
+        this.unregisterNotificationHandler();
+        this._currentHandler = handler;
```

Comment:
I guess you would generally want to first unregister from this._currentHandler if it already exists.  I think you will also want to keep a record of which listeners have been engaged so that you can reregister them in cases where the connection has been lost and you have reconnected ( see the new instruction for auto reconnect: https://docs.microsoft.com/en-us/aspnet/core/signalr/javascript-client?view=aspnetcore-3.1#reconnect-clients).
Another possible problem I see is that if you register for, e.g., connectNotificationHandler with this method, then you register for some other listener in another method with a new NotificationHandler, your `this._utils.disconnectNotificationHandler(this._currentHandler)` below will fail, since this._currentHandler has been redefined and `this._currentHandler.handleUpdatedEdition` no longer refers to the `handler.handleUpdatedEdition` with which the "on" listener was created, so the listener will not be unregistered (the "on" and "off" must reference the same function, not just a function that does the same thing).

Reply:
This is the older SignalR code. Look at the signalr branch for the new code.

I did not implement auto reconnection yet, I need to figure out how to do that, or even if there's any point in implementing it at all. If the connection is down and then up again, we may miss notifications in the meantime, which means we need to refresh whatever the user is currently viewing. I can add a mechanism to keep track of this (I need to somehow mark certain prepare functions as those that needs that need to be recalled after reconnecting), or we can just show a "Please refresh this page" notification to the user, much like Facebook does.

I suggest we try with the refresh message, which is very easy to implement, and add automatic recovery only if it bothers users in the real world.

Reply:
Sorry, I just took a look at the other branch and that makes sense now.
I agree with you about reconnect, it is probably too fragile to do anything other than what you suggest.